Improve Handling of Output Produced by Core Executables When Invoked By Services

dmylov opened the pull request:
https://github.com/ngageoint/hootenanny/issues/1342

dmylov commented:
@brianhatchl, lots to review.  I've practically touched every file in the Services.

dmylov commented:
run cucumber

sputnam commented:
run cucumber

dmylov commented:
@brianhatchl Everything is ready for your review.

dmylov commented:
run cucumber

sputnam commented:
run copyright

brianhatchl commented on hoot-services/src/main/java/hoot/services/HootProperties.java:
```diff
@@ -228,15 +214,26 @@
+
+        // Adding another layer of indirection for "sensitive" properties.
+        // They should be resolved just before being used to minimize any unintended exposure (f.e. logging).
+        DB_NAME = "${dbName}";
```
This is a good change.

brianhatchl commented on hoot-services/src/main/java/hoot/services/command/ExternalCommandRunnerImpl.java:
```diff
@@ -0,0 +1,173 @@
+            commandResult.setFinish(finish);
+            commandResult.setJobId(jobId);
+
+            updateDatabase(commandResult);
```
So we're writing job status to the db including stdout.  Does this table ever get truncated?  Seems like this could grow pretty large over time.

dmylov replied:
Right now, we don't have any logic to truncate/prune the table when it grows too large in size.  We can always add a custom service within Services to always keep the table's size in check by leaving  only the last n command_status records.  This can also be a DB maintenance step by someone in charge of a Hootenanny instance.

sputnam commented:
run copyright

dmylov commented on hoot-services/src/main/java/hoot/services/command/ExternalCommandManagerImpl.java:
```diff
@@ -0,0 +1,177 @@
+        }
+        else if (exec.equalsIgnoreCase("bash")) {
+            commandArr = createBashScriptCmdArray(command);
+        }
```
This if-else statement that checks for 'make', 'hoot', and 'bash' has been there for some time.  Once we decide on a more consistent approach of interfacing with the Core, I anticipate this code to be refactored.

dmylov commented:
List of changes included into this pull request:

1) Added command_status table
2) Major code repackaging to promote code organization and cohesiveness  
3) Removed all job processing calls that used reflection
4) Defined what Command and Job are 
5) Moved as much command related logic out of the Web controllers into their own classes as possible
6) Classified commands into external and internal commands
7) Enhanced the code to implement special handling of "sensitive" properties
8) Refactored significant number of legacy unit tests and added new ones
9) Increased the code's modularization
10) Improved the code's testability
11) Performed general code cleanup
